[Feature] 검색결과리스트화면 및 여행지상세화면 UI 추가 구현

letusHyun opened the pull request:
🚨 **Your checklist for this pull request** 

- [x]  Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!

- [x]  Check the commit's or even all commits' message styles matches our requested structure.

## 📌 [요약]
- CopyAlertView 구현
	- Common디렉토리 -> View디렉토리 -> CopyAlertView.swift 추가
- 여행지상세화면 UI 구현
	- 여행지상세화면 복사버튼 클릭 시, 클립보드에 text 복사
- 메인 검색 탭 화면(SearchViewController)에서 searchView를 내비바로 변경
	- 내비바 클릭 시, UI 변경
- 검색 결과 리스트 화면(SearchResultListViewController) viewModel, coordinator, input 및 state 구현


## 📸  [스크린샷(선택)]
<img src="https://github.com/IF-TG/iOS/assets/108918481/183157ec-6fd6-40ca-a34f-1df41179085a" width="350">|<img src="https://github.com/IF-TG/iOS/assets/108918481/b484aa48-5596-4dad-9c53-c78729456f7f" width="350">
|:-:|:-:|
|`여행지 상세 화면`|`검색 결과 화면`|

## TODO
- 여행지상세화면 스크롤에 따른 이미지뷰 UI 변경하기
- content를 담당하는 section의 decorationItem radius 변경하기
	- reuseableView 한계 해결하기..
- 각 화면에 대한 UseCase, Repository 구현하기
- 애플로그인기능 완성하기
- 검색 결과 리스트 화면에서 카테고리 cell 클릭 시, 적정 스크롤링을 통해 UX 개선하기

## 📚 [레퍼런스 (또는 새로 알게 된 내용) 혹은 궁금한 사항]
- 클립보드에 텍스트 복사하기
	- https://developer.apple.com/documentation/uikit/uipasteboard

letusHyun committed:
[Add] SearchResultListCoordinator 추가

letusHyun committed:
[Refactor] PostSearchVC, VM 구조 리펙토링

letusHyun committed:
[Feat] PostSearchCoordinator에 SearchType 정의

letusHyun committed:
[Feat] SearchCoordinator에서 showPostSearch함수 구현

letusHyun committed:
[Feat] PostSearchActions 추가로 인한 코드 수정

letusHyun committed:
[Feat] searchType에 따라서 화면 전환 제어

letusHyun committed:
[Style] 함수 네이밍 변경

letusHyun committed:
[Feat] SearhResultList화면의 input, output, actions, coordinator 구현

letusHyun committed:
[Style] PostSearch화면의 state 네이밍 변경

letusHyun committed:
[Refactor] 메인 Search화면에 navigation bar 구현

letusHyun committed:
[Feat] 검색 화면의 타입을 구분해주는 SearchType 구현

letusHyun committed:
[System] pbxproj

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into refactor-searchBar-to-naviItem

letusHyun committed:
[Style] post search headerview의텍스트 bold 수정

letusHyun committed:
[Feat] Search 탭에서 내비게이션 바 반투명도 제거

letusHyun committed:
[Feat] SearchVM 주입 및 action 추가

letusHyun committed:
[Refector] SearchCoodinatorDelegate 제거

letusHyun committed:
[Style] 내비바 appearance 수정

letusHyun committed:
[Feat] 데코뷰 사용해서 레이아웃 패딩 값 적용

letusHyun committed:
[Refactor] SearchDestination화면 modeling 수정

letusHyun committed:
[Refactor] 이 cell도 모델링 같이 수정

letusHyun committed:
[Feat] SearchDestinationViewController에 viewModel 적용

letusHyun committed:
[Temp] 임시 코드

letusHyun committed:
[Feat] SearchDestinationTitleCell의 shadow 적용

letusHyun committed:
[Style] mock 변경

letusHyun committed:
[Chore] package

letusHyun committed:
[Feat] 여행지화면에서 이미지뷰 상단에 shadow 적용

letusHyun committed:
[Refactor] titleCell에 존재하는 복사버튼을 combine 사용하는 것으로 수정

letusHyun committed:
[Style] SearchDestinationCompositionalLayout 0번째 section top padding값 변경

letusHyun committed:
[Feat] 탭바의 visibility를 담당하는 메소드를 UIViewController extension에 추가

letusHyun committed:
[Comment] 주석 추가

letusHyun committed:
[Style] common 메소드 적용

letusHyun committed:
[Feat] CopyAlertView 구현

letusHyun committed:
[Feat] 타이틀 cell에서 복사 버튼 클릭 시, 클립보드에 주소text 저장

letusHyun committed:
[Feat] copyAddressView 애니메이션 적용

letusHyun committed:
[Feat] appearCopyAlert state 적용

letusHyun committed:
[System] pbxproj

letusHyun committed:
[Style] temp image 변경

letusHyun committed:
[Feat] TravelDestinationInfo 프로퍼티 추가

letusHyun committed:
[Style] TravelDestinationInfo 프로퍼티 추가에 따른 mock 변경

letusHyun committed:
[Feat] SearchResultCategoryCell에 categoryId 추가

letusHyun committed:
[Feat] SearchResultList화면에서 카테고리 선택 시 input 정의

letusHyun committed:
[Feat] SearchResultListVC의 reloadSection state 추가

letusHyun committed:
[Feat] TravelDestinationCategoryInfo 추가

letusHyun committed:
[Feat] TravelDestinationCell에 contentId 저장프로퍼티 추가

letusHyun committed:
[Feat] TravelDestinationCell에 bind함수 추가

letusHyun committed:
[Feat] SearchResultListVM input에 대한 output 구현

letusHyun committed:
[System] pbxproj

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into refactor-searchBar-to-naviItem

# Conflicts:
#	travelPlan/Source/Presentation/Search/ViewModel/SearchResultListViewModel.swift
#	travelPlan/travelPlan.xcodeproj/project.pbxproj
#	travelPlan/travelPlan.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

letusHyun committed:
[Chore] package

letusHyun committed:
[Merge] merge 추가

SHcommit commented on travelPlan/Source/Common/Extension/UIKit+/UIViewController+.swift:
```diff
@@ -13,4 +13,15 @@ extension UIViewController {
   }
+  
+  /// 탭바의 Visibility를 담당합니다.
+  func updateTabBarVisibility(_ visible: Bool) {
```
오홍 좋은 기능이 추가되었네요: )

SHcommit commented on travelPlan/Source/Common/View/CopyAlertView.swift:
```diff
@@ -0,0 +1,51 @@
+import UIKit
+import SnapKit
+
+final class CopyAlertView: UIView {
```
다른 화면에서도 활용되는 것 같습니다..  UIVC+에 추가해서 함수 호출시 보여졌다 사라지는 것을 담당하는 함수 추가하는 것도 좋을것 같습니다. 

- 좀 더 익숙한 네이밍으로 토스트바로 네이밍하는 방법 어떨까욧?.?

letusHyun replied:
오호 더 좋은 것 같습니다. 다만, 애니메이션처리를 스르륵 보여졌다가 사라지게 하는 애니메이션 처리로 구현하고싶은데 토스트네이밍보다 고스트(유령)뷰 네이밍은 어떨까요?
ex) PrimaryColorGhostView

SHcommit commented on travelPlan/Source/Presentation/Search/View/SearchResultCategoryCell.swift:
```diff
@@ -10,12 +10,16 @@ import SnapKit
 final class SearchResultCategoryCell: UICollectionViewCell {
   // MARK: - Properties
-  static let id = String(describing: SearchResultCategoryCell.self)
+  static var id: String {
+    return String(describing: SearchResultCategoryCell.self)
+  }
```
앗,, CellIdentifiable을 준수하도록 추가를 안했네요.

```
extension UICollectionViewCell: CellIdentifiable {}
extension UITableViewCell: CellIdentifiable {}
```

이 두개를 추가한다면 위 선언 코드 대신 identifier를 호출해도 좋을것 같습니다.

letusHyun replied:
수정하겠습니다!

SHcommit commented on travelPlan/Source/Presentation/Feed/ViewModel/PostSearchViewModel.swift:
```diff
@@ -104,45 +94,47 @@ extension PostSearchViewModel: ViewModelCase {
     return input.didTapSearchButton
-      .tryMap { State.gotoSearch(searchText: $0) }
-      .mapError { $0 as? ErrorType ?? .unexpected }
+      .map { [weak self] text in
+        guard let self = self else { return State.none }
+        switch searchType {
+        case .travelDestination:
+          actions.showTravelDestinationList(text)
+        case .post:
+          actions.showPostList(text)
+        }
+        return State.none
+      }
```
actions 호출 -> 화면 전환 후에 State. none이 이루어질 것 같습니다.
화면 전환시에 VC의 코디네이터를 이용하지 않는다면 Input -> Otuput 으로 뷰한테 상태를 전달하지 않아도 좋을것 같습니다.

ViewModel에서 페이지 전환 Delegate protocol를 추가 후 actions를 호출해 화면 전환하도록 준수하고, 뷰 컨트롤러에서는 이 프로토콜 함수를 호출해서 바로 메인 스레드에서 화면 전환이 호출될 수 있도록 하는 방법은 어떨까요? 

이렇게 할 경우 코드가 늘어나게 됩니다.

letusHyun replied:
State.none이 마음에 걸렸는데, delegate 패턴을 사용하면 불필요한 코드가 줄어들어서 좋은 것 같습니다!

SHcommit reviewed (approved):
오홋,, 확인했습니다: )

--- 
~~### [ 검색 상세 화면에 상단 노치부분 ]~~

<img src="https://github.com/IF-TG/iOS/assets/96910404/c88b5938-11f9-440e-9740-045765f11678" width="250">|
|:-:|
| `노치 inset조정 `|

~~- 이중 스크롤로 구현하는 방법~~
~~- 스택뷰에 이미지뷰 + 컨텐츠뷰~~
~~- 컬랙션뷰~~

~~상의했던 이 세가지 중에서 컬랙션뷰를 활용해봤습니다.~~

~~검색 상세 화면에 상단 노치부분을 좀 구현해봤는데 section내 inset 조정은 해당 섹션의 group 내 cell들은 상단으로 inset 적용이 가능하지만, reusalbe view를 section의 헤더로 사용했을때, section's header는 inset이 적용이 안됬습니다. 헤더에 자체적인 inset을 적용했어도 마찬가지입니다.😅~~

~~노치부분을 그냥 다른 section의 cell로 만들어서 적용하니 section의 인셋을 주어 위 영상처럼 적용 됩니다.~~

~~이렇게 구현할 때 주의해야 할 부분은 dequeueReusable로 꺼내올 때 맨 아래 스크롤 후, 맨 위로 스크롤할 때, 노치 cell 나온 후 gray section의 cell이 화면에 보여지게 됩니다.~~

<img src="https://github.com/IF-TG/iOS/assets/96910404/05335b7b-bb05-44fb-acf5-5265db2d4666" width="250">|
|:-:|
|`reusable view 최근에 꺼내진 뷰의 계층이 상단에 위치! `|

~~컬랙션뷰, 컬랙션 뷰 특성상 맨 나중에 꺼내진 셀이 맨위로 보여지며 뷰 계층구조가 이루어집니다. 이때 맨 나중에 꺼내진 셀이 section index == 0인 경우 현제 화면에 보이는 노치셀을 제일 앞으로 꺼내오면 해결됩니다.~~

~~section 0 == 이미지~~
~~section 1 == 노치부분~~
~~section 2 == 컨텐츠 부분~~

~~🥹~~

letusHyun committed:
[Refactor] cellIdentifier, HeaderFooterIdentifiable 을 각각 사용하는 객체가 extension으로 준수하도록 수정

letusHyun committed:
[Refactor] CellIdentifiable 최종 수정

letusHyun committed:
[Chore] package

letusHyun committed:
[Rename] CopyAlertView를 LandscapeToastView로 네이밍 변경

Landscape이라는 용어가 가로모드 혹은 비율이 세로보다 가로가 더 길 때 사용한다고 해서 LandscapeToastView로 네이밍을 지어봤습니다.

letusHyun committed:
[Style] 토스트뷰 네이밍 변경에 따른 코드 수정

letusHyun committed:
[Feat] init에 color, text를 매개변수로 설정

letusHyun committed:
[Style] 불필요 코드 제거

letusHyun committed:
[Chore] package

letusHyun committed:
[Refactor] 페이지 전환 시작점인 input 제거하고, delegate패턴 적용

letusHyun committed:
Merge branch 'develop' of https://github.com/IF-TG/iOS into refactor-searchBar-to-naviItem

# Conflicts:
#	travelPlan/Source/Presentation/Feed/Coordinator/FeedCoordinator.swift
